### PR TITLE
gcoap: add DTLS integration

### PR DESCRIPTION
**Pull request by pokgak**
### Contribution description

This PR add DTLS support in gcoap.

Among the changes done in this PR are:
  - Add a new function `gcoap_set_credential_tag()` to tell [credman][1] which credential to use. For other application that is similar to gcoap where the user does not have direct access to the sock used, this is how DTLS integration using [DTLS sock API][6] can be done.
  - Provide top level transport layer (96e3a68) and use common name for the sock (6c8f9cd) to abstracts away the transport mechanism from gcoap logic. This is contributed by @kb2ma. More on that in pokgak/RIOT#1.
  - Example of using DTLS in a gcoap application.

[Section 9.1 of the CoAP RFC][7] mentions checks that must be done for CoAP implementations that integrates DTLS. I have not done any of said checks in this PR yet.

### Testing procedure

NOTE: This PR will currently fail compile due to dependencies on other PR. For testing, you will need to manually pull in changes from #11943 and #12529.

Testing have been done between two `samr21-xpro` and also between a `nucleo-l476rg` and a RPI with 802.15.4 radio. On the RPI, I've tested this PR with the following stack on the Pi:
  - [libcoap + tinydtls][2] as server and client using the example applications `coap-server` and `coap-client`
  - [aiocoap + tinydtls][3] as client using `aiocoap-client`
  - [coap-shell][4] as client which uses californium + scandium

For testing between two RIOT nodes, simply flash `example/gcoap` onto the boards and send a CoAP get from the client to the server using:

```
coap get <server ip> 5684 /.well-known/core
```

When testing with a Pi, I encountered some problem when using link-local IPv6 so it is better to use the global address. For that some changes are needed to the Makefile:

```diff
-USEMODULE += gnrc_ipv6_default
+USEMODULE += gnrc_ipv6_router_default

...

+CFLAGS += -DGNRC_IPV6_NIB_CONF_SLAAC=1
```

Then [setup RADVD][5] on the Pi and you can proceed with the instructions above.

### Issues/PRs references

DTLS sock tracking issue: #10897

[1]: http://api.riot-os.org/group__net__credman.html
[2]: https://github.com/obgm/libcoap
[3]: https://github.com/chrysn/aiocoap
[4]: https://github.com/tzolov/coap-shell
[5]: https://github.com/RIOT-Makers/wpan-raspbian/wiki/Setup-native-6LoWPAN-router-using-Raspbian-and-RADVD
[6]: http://api.riot-os.org/group__net__sock__dtls.html
[7]: https://tools.ietf.org/html/rfc7252#section-9.1

### CODE REVIEW

**Comment by pokgak** [comment]
Pinging people that I think might be interested in this @kb2ma @leandrolanzieri @smlng @tcschmidt @chrysn

**Comment by benpicco** [comment]
What's missing for this to not be WIP anymore?

**Comment by kb2ma** [comment]
Thanks for working to move this forward, @benpicco. For my part I can only say it is high on my list to review after 2019.10 is released.

**Comment by tcschmidt** [comment]
@benpicco I guess you're right, it's not WIP anymore. It just waits for the review ... and we all understand why :-)

**Comment by benpicco** [comment]
Oh I don't know why - I thought I'd wait until this is not WIP anymore before beginning a review.

**Comment by benpicco** [comment]
@pokgak  If this is not WIP anymore, please squash!
Review hasn't started yet and it's much easier to review individual commits.

**Comment by smlng** [comment]
@pokgak please clarify: is  #11943 needed for this or not? Bc if it is, we should add the dependency and "waiting for PR" label, too. Also it makes more sense to merge  #11943 first.

**Comment by pokgak** [comment]
Thanks for the interest!

> What's missing for this to not be WIP anymore?

Some checks as required by the RFC is still not done yet.

> [Section 9.1 of the CoAP RFC](https://tools.ietf.org/html/rfc7252#section-9.1) mentions checks that must be done for CoAP implementations that integrates DTLS. I have not done any of said checks in this PR yet.

These needs to be move to another PR:

> * Move `gcoap_init()` out of `auto_init` to first load the credentials before start listening for CoAP requests. It is called manually from user application (here: `gcoap_cli.c`)
> 
> * Add a new function `gcoap_set_credential_tag()` to tell [credman](http://api.riot-os.org/group__net__credman.html) which credential to use. For other application that is similar to gcoap where the user does not have direct access to the sock used, this is how DTLS integration using [DTLS sock API](http://api.riot-os.org/group__net__sock__dtls.html) can be done.

Other than that, I have an incomplete version of DTLS sock with WolfSSL (currently this PR is only using tinydtls with DTLS sock), , that I'm using to make sure that the changes I've done here is not specific to tinydtls. For example setting the `sock_dtls_session_t` currently is specific to tinydtls:

https://github.com/RIOT-OS/RIOT/blob/8a2f70adf447101651b6b0885ca80ac70d392caa/sys/net/application_layer/gcoap/gcoap.c#L118-L123

A function to wrap that will need to be added to the DTLS sock API. Added this to TODO in tracking issue.

> @pokgak please clarify: is #11943 needed for this or not? Bc if it is, we should add the dependency and "waiting for PR" label, too. Also it makes more sense to merge #11943 first.

I prefer #11943 gets merged first before review on this PR, so you could set the "waiting for PR" label for now.

**Comment by pokgak** [comment]
I updated this PR and it is now not a WIP anymore. Compiling will still fail because of dependencies on other PRs (#11943, #12529).

> [Section 9.1 of the CoAP RFC](https://tools.ietf.org/html/rfc7252#section-9.1) mentions checks that must be done for CoAP implementations that integrate DTLS. I have not done any of the said checks in this PR yet.

I skipped this check based on the response I got in https://github.com/obgm/libcoap/issues/420#issuecomment-550320787.

> Other than that, I have an incomplete version of DTLS sock with WolfSSL (currently this PR is only using tinydtls with DTLS sock), , that I'm using to make sure that the changes I've done here are not specific to tinydtls. For example setting the `sock_dtls_session_t` currently is specific to tinydtls:
> 
> https://github.com/RIOT-OS/RIOT/blob/8a2f70adf447101651b6b0885ca80ac70d392caa/sys/net/application_layer/gcoap/gcoap.c#L118-L123

This can be done in follow up PR for DTLS sock. I prefer not to change that PR anymore until it is merged.

TLDR: this PR is ready for review and testing it as is is not possible but I tested it locally by pulling in the dependencies from #11943 and #12529.

**Comment by benpicco** [comment]
So this needs a rebase now 😃

**Comment by pokgak** [comment]
Rebased to master and tested working with `native`.

**Comment by kb2ma** [comment]
Thanks @pokgak, I will take a look in the next few days.

**Comment by miri64** [comment]
I'm not sure, but don't there need to be some URLs also to be changed when DTLS is used: `coap://` -> `coaps://`

**Comment by miri64** [comment]
And ports. Otherwise clients might have trouble communicating with `gcoap`.

**Comment by miri64** [comment]
Also, I'm not sure we want to hard-code the protocol at compile-time. There are legitimate use-cases for providing and requesting both `coap://` and `coaps://` resources at the same time.

**Comment by pokgak** [comment]
> I'm not sure, but don't there need to be some URLs also to be changed when DTLS is used: `coap://` -> `coaps://`?

AFAIK currently the gcoap API doesn't need to know the if the URL is prefixed `coap://`. `gcoap_req_send()` for example expects a `sock_udp_ep_t`. The information in that sock object must be filled by users. So based on that I would say if there is a need to differentiate URL between `coap://` and `coaps://` then it is left to the users to implement the check.


> And ports. Otherwise clients might have trouble communicating with `gcoap`.

The port `5684` is used automatically when DTLS is used. When using the gcoap example application, running `coap info` will also show which port the server is listening on.

https://github.com/RIOT-OS/RIOT/blob/84d99ac6c2bcdc6225870450a65d04b1f50e2015/sys/include/net/gcoap.h#L379-L385

> Also, I'm not sure we want to hard-code the protocol at compile-time. There are legitimate use-cases for providing and requesting both `coap://` and `coaps://` resources at the same time.

I see your point and that is also how other CoAP libraries (Californium, libcoap) do it. Using one thread per protocol (CoAP over {UDP, DTLS} and later maybe also TCP, TLS)  might be a way to do this but that may involve bigger changes to the existing code and more memory is needed due to multiple threads. Any thoughts on this @kb2ma?

**Comment by miri64** [comment]
> I see your point and that is also how other CoAP libraries (Californium, libcoap) do it. Using one thread per protocol (CoAP over {UDP, DTLS} and later maybe also TCP, TLS) might be a way to do this but that may involve bigger changes to the existing code and more memory is needed due to multiple threads. Any thoughts on this @kb2ma?

With `sock_async` you wouldn't need an additional thread, just some (de-)multiplexing between the UDP and DTLS sock ;-) (need to kink out the compilation issue in #12907 first though :sweat_smile:).

**Comment by pokgak** [comment]
Yeah, that would be the alternative. I'll take a look at #12907 then and see how it fits.

**Comment by kb2ma** [comment]
I also initially assumed multiple protocols/ports would mean multiple threads. If there is a possibility to multiplex on a single thread, then definitely let's investigate that.

**Comment by kb2ma** [comment]
I just started to look at the async stuff. Do we really want to try to bring that in on this PR? As someone unfamiliar with that work, I already need to review #12625 and #12602 to see how they affect gcoap. 

Without knowing the details, I am in favor of keeping the compile time selection of protocol for now, to take a smaller step that focuses on gcoap/DTLS integration. @miri64? @pokgak?

**Comment by miri64** [comment]
> I just started to look at the async stuff. Do we really want to try to bring that in on this PR? As someone unfamiliar with that work, I already need to review #12625 and #12602 to see how they affect gcoap.
> 
> Without knowing the details, I am in favor of keeping the compile time selection of protocol for now, to take a smaller step that focuses on gcoap/DTLS integration. @miri64? @pokgak?

If you think that is the better approach for now than we should do it like this. We'd need to adapt `gcoap` for `sock_async` anyway to fix #8130.

**Comment by kb2ma** [comment]
Thanks, @miri64. I reviewed the async work, and it looks well suited to gcoap's use. However, I think it makes more sense to hold off on integration until after this PR.

It was useful to imagine how gcoap would need to adapt to async. It will need to maintain a udp sock and a dtls sock. Also, a request must specify which sock to use. I think it will make sense to rework `gcoap_req_send()` to accept a new struct parameter, something like `gcoap_request_options_t`. This struct will include the new user `context` attribute from #9857. It also will include some identifier for the sock to use. I'm also quite sure we will find additional uses for it in the future. ;-)

The `_tl_send()` function in this PR will need to change for async. But I think it could be adapted with a a union based parameter for a udp sock or a dtls sock. Then there would still be only a single function/location in gcoap with conditional logic to use the proper sock send function.

**Comment by pokgak** [comment]
> The `_tl_send()` function in this PR will need to change for async. But I think it could be adapted with a a union based parameter for a udp sock or a dtls sock. Then there would still be only a single function/location in gcoap with conditional logic to use the proper sock send function.

Should I try to do this for this PR?

**Comment by kb2ma** [comment]
> Should I try to do this for this PR?

No, sir! `_tl_send()` is internal to gcoap.c, so we can change it in the future. I was just speculating on how we might adapt to the async work.

My next step is some functional testing.

File: examples/gcoap/gcoap_cli.c [hunk 1]
```diff
@@ -26,8 +26,24 @@
...
+extern const unsigned psk_key_len;
+extern const unsigned psk_id_len;
+#else /* DTLS_PSK */
+extern const unsigned char ecdsa_priv_key[];
+extern const unsigned char ecdsa_pub_key_x[];
+extern const unsigned char ecdsa_pub_key_y[];
+#endif /* DTLS_ECC */
+#endif /* MODULE_SOCK_DTLS */
+
+#define ENABLE_DEBUG (1)
```
[the remark this answers is not in the document]
Comment (pokgak) [reply examples/gcoap/gcoap_cli.c]: Oopsie.

File: sys/Makefile.dep [hunk 2]
```diff
@@ -19,4 +19,13 @@ ifneq (,$(filter prng_fortuna,$(USEMODULE)))
   CFLAGS += -DCRYPTO_AES
 endif
 
+ifneq (,$(filter gcoap,$(USEMODULE)))
+  ifneq (,$(filter sock_dtls,$(USEMODULE)))
+    CFLAGS += -DGCOAP_NO_AUTO_INIT
+    # Payload size might be bigger when using DTLS with ECC, causing errors not present
+    # when using plain CoAP
+    CFLAGS += -DGCOAP_PDU_BUF_SIZE=256
```
[the remark this answers is not in the document]
Comment (pokgak) [reply sys/Makefile.dep]: yeah you're right. I'll move this to `gcoap.h` and add the comments to the docs.

File: sys/include/net/gcoap.h [hunk 3]
```diff
@@ -602,6 +621,13 @@ extern "C" {
 #define GCOAP_RESEND_BUFS_MAX      (1)
 #endif
 
+/* define a common name for transport layer object, to accommodate security layer */
+#ifdef MODULE_SOCK_DTLS
+typedef sock_dtls_t coap_tl_sock_t;
+#else
+typedef sock_udp_t coap_tl_sock_t;
+#endif
```
Comment (kb2ma) [inline sys/include/net/gcoap.h]: Doxygen docs complain about absent documentation for these definitions. Since they are used only in _listen() in `gcoap.c`, you can move these definitions there.

Comment (pokgak) [reply sys/include/net/gcoap.h]: Moved to `gcoap.c` and also added usage instruction to docs in c52b34b.

File: sys/net/application_layer/gcoap/gcoap.c [hunk 4]
```diff
@@ -42,7 +42,7 @@
 
 /* Internal functions */
 static void *_event_loop(void *arg);
-static void _listen(sock_udp_t *sock);
+static void _listen(coap_tl_sock_t *sock);
```
Comment (kb2ma) [inline sys/net/application_layer/gcoap/gcoap.c]: This declaration must be below the definition of `coap_tl_sock_t`.

File: examples/gcoap/credentials.c [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+
+/**
+ * @ingroup     tests
+ * @{
+ *
+ * @file
+ * @brief       tlsman test application (PSK and ECC keys)
```
Comment (cgundogan) [inline examples/gcoap/credentials.c]: The `tlsman` module of @rfuentess was never merged. We should change the wording here to not confuse newcomers.

Comment (pokgak) [reply examples/gcoap/credentials.c]: Edited the @brief and description.

File: sys/net/application_layer/gcoap/gcoap.c (hunk 4 above)
Comment (pokgak) [reply sys/net/application_layer/gcoap/gcoap.c]: Fixed this in https://github.com/RIOT-OS/RIOT/pull/12104/commits/236732f3224bbc7aeabf301b92d94aed7e2354ce

File: sys/include/net/gcoap.h [hunk 6]
```diff
@@ -373,15 +400,23 @@ extern "C" {
...
+#ifdef  MODULE_SOCK_DTLS
+#define GCOAP_PORT              (5684)
+#else
 #define GCOAP_PORT              (5683)
-#endif
+#endif  /* MODULE_SOCK_DTLS */
+#endif  /* GCOAP_PORT */
 
 /**
  * @brief   Size of the buffer used to build a CoAP request or response
```
Comment (kb2ma) [inline sys/include/net/gcoap.h]: I expected to see a comment similar to what was in `Makefile.dep`. The additional 128 bytes deserves some explanation so the user can change it if that is reasonable.

> Payload size might be bigger when using DTLS with ECC, causing errors not present
> when using plain CoAP

File: examples/gcoap/gcoap_cli.c [hunk 7]
```diff
@@ -26,6 +28,22 @@
 #include "net/gcoap.h"
 #include "od.h"
 #include "fmt.h"
+#ifdef MODULE_SOCK_DTLS
+#include "net/credman.h"
+
+#define SOCK_DTLS_GCOAP_TAG (10)
```
Comment (kb2ma) [inline examples/gcoap/gcoap_cli.c]: Why did you choose 10 for this value? Why not 1 or 5?

Comment (kb2ma) [inline examples/gcoap/gcoap_cli.c]: More importantly, why define this value here? If it was defined in `gcoap.c` or `gcoap.h`, you would not need `gcoap_set_credential_tag()`, right? Is there any value to defining it in the application? Would gcoap ever need more than one credential? Do we need a distinction between a gcoap server credential and a client credential?

**Comment by kb2ma** [comment]
Functional testing is going OK. I'm updating the automated tests in [riot-coap-test](https://github.com/kb2ma/riot-coap-pytest). It took some work to update my installation of libcoap and aiocoap for DTLS, as you can see in the Materials section of [README.md](https://github.com/kb2ma/riot-coap-pytest/blob/master/README.md). I think it will take another day or two to get through this testing.

File: examples/gcoap/gcoap_cli.c (hunk 7 above)
Comment (kb2ma) [reply examples/gcoap/gcoap_cli.c]: Following this line of thinking, why not move all of the code in `gcoap_cli.c` into gcoap itself? gcoap provides CoAP as a service, so why should every application need to add this code? With this approach the application/user only needs to provide `credentials.c`.

Comment (pokgak) [reply examples/gcoap/gcoap_cli.c]: > More importantly, why define this value here? If it was defined in `gcoap.c` or `gcoap.h`, you would not need `gcoap_set_credential_tag()`, right? Is there any value to defining it in the application?

Yup, you're right. The credential tag is defined in the client to avoid reserving values for the tag used in gcoap. Other modules (e.g. LWM2M) might also hard-code the tag used for the DTLS sock and then there will be a need to track the available tag somewhere to avoid collisions.

> Would gcoap ever need more than one credential?

Yes. gcoap might be talking to other secure coap clients/servers that use different credentials.

> Do we need a distinction between a gcoap server credential and a client credential?

No.

> Following this line of thinking, why not move all of the code from this PR in `gcoap_cli.c` into gcoap itself? gcoap provides CoAP as a service, so why should every application need to add this code? With this approach, the application/user only needs to provide `credentials.c`.

The credentials are added in `gcoap_cli.c` because we do not want to make any assumption or limit what the name of the variables used for the credentials should be. 

In general, the design of coupling DTLS sock with credman is to avoid limiting what the name of the credentials should be or where it can be found. This is left to the user to define.

Comment (pokgak) [reply examples/gcoap/gcoap_cli.c]: This can be any `unsigned` values. No reason specifically why 10 is chosen here.

File: sys/include/net/gcoap.h (hunk 6 above)
Comment (pokgak) [reply sys/include/net/gcoap.h]: Rephrased and added the comment in dfdbb86c0

**Comment by kb2ma** [comment]
I'd like to be sure this PR supports use of more than one credential. So, suppose there are two credentials, C1 and C2, with tags 1 and 2 respectively. C1 is the credential the gcoap CLI uses to initialize gcoap. Let's assume SOCK_DTLS_GCOAP_TAG is 1.

An endpoint A sends a GET request to the gcoap device using credential C1. gcoap's `_tl_sock` of type `sock_dtls_t` receives the request, the credential matches, and gcoap sends the response to endpoint A.

Now suppose I use the gcoap CLI to send a message to an endpoint B. However, endpoint B requires use of credential C2. The CLI uses gcoap_req_send() to send the message to B. However, `gcoap_req_send()` calls the internal `_tl_send()` function, which uses gcoap's _tl_sock, which uses credential C1. This request will fail then, won't it?

Let's assume we find a way for gcoap to send the message to endpoint B with a different sock. Now gcoap also needs to listen for a response on this sock for credential C2 as well as listen on the sock for credential C1. I don't see how this is possible without something like the async mechanism we have been discussing.

Without DTLS, gcoap acts as a messaging hub. All communication can pass through the single `sock_udp_t`sock. It seems like there needs to be some multiplexing somewhere for gcoap to associate the appropriate `sock_dtls_t` and credential with a given request/response.

Am I missing something here?

**Comment by pokgak** [comment]
> Now suppose I use the gcoap CLI to send a message to an endpoint B. However, endpoint B requires use of credential C2. The CLI uses gcoap_req_send() to send the message to B. However, `gcoap_req_send()` calls the internal `_tl_send()` function, which uses gcoap's _tl_sock, which uses credential C1. This request will fail then, won't it?

Yes, sending a message to endpoint B, which uses credential C2 will fail.

---------------------
The above part should answer your question. The following is an idea that I had in the earlier implementation of DTLS sock.

One approach to support use of more than one credential is to use a different sock for each credential as you've described above. Another approach to support more than one credential is to modify how a DTLS sock gets its credentials from credman (the credential storage module).

The reason why a DTLS sock only supports one credential is that each sock can only have a single credential tag. See `sock_dtls::tag` below:

https://github.com/RIOT-OS/RIOT/blob/cfeb2a38431d145070c914aed63fe9d08f6c57b1/pkg/tinydtls/include/sock_dtls_types.h#L37-L50

My early implementation actually uses an array of `credman_tag_t` instead of just a single `credman_tag_t`:

```C
/* current implementation */
struct sock_dtls {
    ...
    credman_tag_t tag;
    ...
};

/* earlier implementation */
struct sock_dtls {
    ...
    credman_tag_t tag[MAX_CREDMAN_TAG];
    ...
} 
```

With this approach, supporting more than one credential per sock can be done by adding a new tag to the array. When looking for which credential to use, the DTLS sock can go through the tag array and find a credential that matches.

The problem with this approach is that going through the tag array each time can be costly if there are a lot of credential tags added to the sock. There is also the problem of how to know which credential tag to use to when a packet arrived. But this approach is just an idea that I had implemented half-way before giving up due to said (possible) problems and opted for a simple and limited current implementation (which proves to be problematic now)

TLDR: The current DTLS sock implementation is limited to only one credential per sock due to the design choices of DTLS sock.

**Comment by kb2ma** [comment]
Thanks for the explanation.

> My early implementation actually uses an array of credman_tag_t instead of just a single credman_tag_t

gcoap certainly would appreciate the ability to continue to use a single sock. The limitation of a single credential significantly affects the usefulness of gcoap. If you think it's reasonable to extend dtls_sock in this way in the near future, then I won't get too concerned over the present limitation, and we can move forward. Please add a warning to the DTLS section of the gcoap doc though.

> The problem with this approach is that going through the tag array each time can be costly if there are a lot of credential tags added to the sock.

If an application needs a lot of credential tags, they need the resources to deal with it. I would think that most applications would have limited requirements here.

> There is also the problem of how to know which credential tag to use to when a packet arrived.

It seems like there is a strong association between a remote endpoint / DTLS session and a credential. dtls_sock could use the remote to find the expected credential.

File: sys/include/net/gcoap.h [hunk 8]
```diff
@@ -305,6 +307,26 @@
...
+ * After that, add credential to use using @ref credman_add(), tell
+ * gcoap which credential tag to use with gcoap_set_credential_tag(). **After**
+ * setting up the credentials, we need to start the gcoap server manually by
+ * calling gcoap_init().
+ *
+ * Sending and receiving CoAP packets with DTLS is the same as using plain CoAP.
+ * gcoap will automatically encrypt/decrypt CoAP packets with DTLS before
+ * sending/receiving them over the network.
+ *
+ * To read more on DTLS support in RIOT, refer @ref net_dtls and @ref net_sock_dtls.
```
Comment (kb2ma) [inline sys/include/net/gcoap.h]: It might be worth specifically mentioning configuration parameters here. I got hung up for a while on the default value of 1 for tinydtls DTLS_PEER_MAX.

**Comment by kb2ma** [comment]
I have finished a first pass at functional testing and updating the riot-coap-pytest automated tests. You can see  what works in the [Setup section](https://github.com/kb2ma/riot-coap-pytest/blob/master/README.md#setup) of the test `README.md`.

There are 3 issues at present:

1. The deadlock addressed in #12959, which causes confirmable retries (con_retry_test) to fail
2. credman limitation of a single credential per sock, described above
3. cord_ep and cord_epsim examples use a synchronous messaging model which fails for DTLS handshaking.

On (3) I was able to manually register with a DTLS based RD server using the gcoap CLI example in PSK mode, but the cord_ep example froze early in handshaking.

**Comment by kb2ma** [comment]
Another issue, as first [described](https://github.com/RIOT-OS/RIOT/pull/12959#issuecomment-571581591) in 12959:

There is no mechanism in gcoap to close a session. I suppose this situation is manageable if gcoap only talks to at most DTLS_PEER_MAX (tinydtls) unique peers. On the other hand, there are scenarios that would like to return a 5.03 Service Unavailable if unable to open a new session.

**Comment by kb2ma** [comment]
@pokgak, @cgundogan -- what's the status with this PR and #12959? Are you blocked or just otherwise busy? Is the best solution to work sock_async into gcoap and #12907 for sock_dtls?

**Comment by pokgak** [comment]
sorry @kb2ma, I'm a stuck with another project right now and won't have much time to work on this for now. 

> Is the best solution to work sock_async into gcoap and #12907 for sock_dtls?

Main problem right now is that sock_dtls blocks while waiting for handshake messages. If we can avoid that with sock_async then it may be the solution.

**Comment by kb2ma** [comment]
> Main problem right now is that sock_dtls blocks while waiting for handshake messages. If we can avoid that with sock_async then it may be the solution.

Gotcha. #13386 adds sock_async and should be ready to experiment with when you get a chance.

**Comment by MichelRottleuthner** [comment]
Just for the record: I played around with this PR a bit and had some problems using it together with various counterparts on the linux side (ie. libcoap, aiocoap, coap-shell).
The guide @kb2ma linked [above](https://github.com/RIOT-OS/RIOT/pull/12104#issuecomment-568308633) then lead me in the right direction. Thanks a lot for the writeup, it was very helpful!.
There still seems to be a problem that is somehow related to the tinydtls version. When I set up tinydtls-cython as described in the linked guide it worked with this PR, but rebasing this PR on top of master broke it again. I didn't dig deeper yet as it was hard to tell if the problems came from tinydtls itself or some other part of the integration. Could be just related to the last version bump (#12912)? Anyway, just let me know when I should give it another shot. I'd really like to see this merged :)

**Comment by kb2ma** [comment]
> Thanks a lot for the writeup, it was very helpful!.

Glad to hear it!

> Anyway, just let me know when I should give it another shot. I'd really like to see this merged :)

If you can verify #12907 sock_async for tinydtls, I can try rebasing this work on that in @pokgak's absence. At least he should be available for questions.

**Comment by MichelRottleuthner** [comment]
> If you can verify #12907 sock_async for tinydtls (...)

my first attempt failed as we don't have a test yet and I'm not sure how to use it properly, but I bet that's solvable with @miri64's help ;)

**Comment by tcschmidt** [comment]
> 
> 
> > If you can verify #12907 sock_async for tinydtls (...)
> 
> my first attempt failed as we don't have a test yet and I'm not sure how to use it properly, but I bet that's solvable with @miri64's help ;)

@pokgak can you help resolving (or fixing) this?

**Comment by pokgak** [comment]
Yup, I'm currently also looking at #12907.

**Comment by pokgak** [comment]
There have been some major changes to gcoap and sock_dtls since this PR is opened. #12907 is working now too. I'm closing this PR and will open another one with the updated version.

My plan is to first rework the credman workflow as per my discussion with @kb2ma and then use that with the new async sock_dtls.

**Comment by benpicco** [comment]
Any news on this?

**Comment by pokgak** [comment]
I'm not working on this at the moment.

**Comment by benpicco** [comment]
Should we take this over?

**Comment by pokgak** [comment]
Yes, go ahead. I think @leandrolanzieri looked into this already, so maybe you guys can sync up.

**Comment by kfessel** [comment]
#15549 did that
